text/0000-startup-initialized-statics.md

```diff
@@ -0,0 +1,97 @@
+
+## A possible implementation
+
+It's getting tedious to write *"startup initialization"* and *"startup initialized"*, so from now on, I'll shorten both of them to SI.
+
+Inside the body of an SI function, all of its associated SI statics are made mutable. It is as if the following kind of transformation happened:
+
+From:
+```rust
+type Data = [i8; 3];
+
+#[startup_initialized(init_my_data)]
+static DATA: Data = [1,2,3];
+
+#[startup_initialization]
+fn init_my_data() {
+    DATA[0] = 11;
+    mutate(&mut DATA);
+}
+
+fn mutate(_: &mut Data) {}
+```
+To:
+```rust
+type Data = [i8; 3];
+
+#[startup_initialized(init_my_data)]
+static DATA: Data = [1,2,3];
+
+#[startup_initialization]
+fn init_my_data() {
+    let tmp: &mut Data = unsafe { std::mem::transmute(&DATA) };
+    (*tmp)[0] = 11;
+    mutate(&mut (*tmp));
+}
+
+fn mutate(_: &mut Data) {}
+```
+Remember that SI statics are never placed in read-only memory, which makes the transmute to `&mut` not be undefined behaviour (I think).
+
+In order to both catch circular dependencies between SI functions and to be able to determine the order in which SI functions should be called at the beginning of main, the compiler must keep track of which SI statics each function and each static item (through a reference) may potentially access. An SI static must not be accessed (neither read nor written to) before its associated SI function is called (and while said SI function call is running, we allow all functions to read and write to the SI static if they take it as a `&mut` argument as long as they don't store a long-lived mutable reference to it, but I think Rust's borrow checking rules prevent this from happening).
```

Comment:
@sfackler I don't really know what non-local analysis means, but just based on how it sounds, it seems to me that the static analysis required here is quite local being pretty well encapsulated to each function, because the analysis relies on the meta-data (dependency sets) that were determined earlier (for the functions being called and statics being accessed) and cached somewhere.